### DISCUSSION TITLE
Can someone explain AIR Native Extensions FREGetContextActionScriptData() and FRESetContextActionScriptData() in more detail?

### DISCUSSION THREAD

**Discussion by manishpatel0404** [airsdk/Adobe-Runtime-Support#978] [Q&A] [answered]
I have created an AIR Native Extension wherein, I have created a reusable Loader object on the AS3 side. I then pass this Loader object to the ANE to load whatever loadable file I want to load into the Loader. This works, as expected.

I would like to create the reusable Loader on the C side of the ANE. My understanding is that the FRESetContextActionScriptData() and FREGetContextActionScriptData() functions allows me to create and store AS3 FREObject variables on the C side.

I have been trying to implement these functions on the C side and and unable to wrap my head around how to do so. The Adobe help is insufficient and does not help in any way. I have tried to search for these two words with no success. Can someone throw some light on how these functions are used?

Thanks in advance.

**Comment by ajwfrost** [comment, marked as answer]
This "ActionScript data" property is something that you can share from AS3 to C via the ExtensionContext object:
https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/external/ExtensionContext.html#actionScriptData

So if you were to set the loader from AS3 you can retrieve it on the C side, and vice versa.

Creating a new object within C and calling methods on it is something that you can do, via the API calls below:
https://help.adobe.com/en_US/air/extensions/WS460ee381960520ad-866f9c112aa6e1ad46-7ff6.html
So it would be possible to create a Loader, call methods/set properties, and then set it on the extension context.

If you explain what you're trying to do, we could look at some sample code perhaps?

thanks

**Reply by manishpatel0404** [reply]
@ajwfrost, thank you for your response.

After having created my first "Hello World!" ANE, I thought of exploring this concept further.  Therefore, I set out to create a toy SWF player ANE. I know it is easy to achieve what the ANE does in AS3 directly, but I am doing this to better understand the process of creating ANEs. Following is what I am trying to achieve:

1. Create a reusable Loader instance and add it to the stage (C Side).
2. Read a swf file into a ByteArray (AS3 side).
3. Set up the event listeners (C Side).
4. Create the loader context (C Side).
5. Load the ByteArray into the Loader instance (C Side).
6. Receive the event (AS3 side).
7. Call the C-side to play the swf content ( (C Side).
8. Repeat steps 2 to 7.

I will then add a UI for the user to select the swf file to be loaded. Sounds silly, but I am learning a lot.

When  I created the Loader instance on the AS3 side and passed it as a parameter to the C-side loadFileBytes and play functions, everything worked perfectly. Then I got ambitious and tried to create the Loader itself on the C-side. I have been struggling ever since.

I have attached the files I am working with. Thanks.
[SampleAne.zip](https://github.com/airsdk/Adobe-Runtime-Support/files/6810803/SampleAne.zip)

**Comment by manishpatel0404** [comment]
Figured out the gaps in my understanding the two functions. Thanks you @ajwfrost for pointing me to the actionScriptData property of the ExtensionContext class, which helped me join the dots. I could not easily figure this out while going through Adobe's AIR Extensions documentation. I have fixed what was wrong in the SampleAne code I had uploaded earlier and it works like a charm.

Again, this is a not a very useful ANE but it helped me learn the internals and that is satisfying.

**Reply by manishpatel0404** [reply]
Just a follow-up question on actionScriptData.

If one needs to store more objects in actionScriptData, would that mean that one would have to create an Array or Associative Array that holds these objects and then assign the Array/Associative Array to the actionScriptData property?

Thanks.

**Reply by ajwfrost** [reply]
Yes, there's only one object reference in actionScriptData so this would need to hold your other objects in some way.

I'd wonder why you need to do this though. Can you pass the objects from ActionScript to C as function arguments in the functions where you need them? Normally you'd have a wrapper ActionScript class that exposes your ANE's API to an application, so this would be able to hold other member variables and any state information, and then make the calls down to C as/when necessary.

**Reply by manishpatel0404** [reply]
Oh! The follow-up question was more in academic interest and to complete the circle of my understanding of actionScriptData.

I agree with you that it is more practical to use a wrapper ActionScript class to expose the ANE's API to the application.

Thanks.
